src/PhpSpec/CodeAnalysis/TokenizedTypeHintRewriter.php

```diff
@@ -198,7 +198,9 @@ private function extractTypehints(array &$tokens, int $index, array $token): voi
 
     private function haveNotReachedEndOfTypeHint($token) : bool
     {
-        if ($token == '|' || $token == '&') {
+        // PHP 8.1 returns the intersection token `&` as an array,
+        // while previous versions return it as a string.
+        if ($token == '|' || $token == '&' || (is_array($token) && $token[1] == '&')) {
```

Comment:
This will break if PHP returns a token-array with only one item. Should I check for it, or can I count on the array to always have a `1`-index? :thinking:

Reply:
I'd say that the explicit check isn't hard to add but might make the code somewhat harder to understand:

```php
if ($token == '|' || $token == '&' || (is_array($token) && ($token[1] ?? '') == '&')) {
    //...
}
```

It could be more readable by separating out the expression into independent booleans:
```php
$isFoo = is_string($token) && $token == '|' || $token == '&';
$isBar = is_array($token) && ($token[1] ?? '') == '&';

if ($isFoo || $isBar) {
    return false;
}
```

Best of luck with naming `$isFoo` and `$isBar` correctly ...